// @cloudinary/url-gen/qualifiers/artisticFilter.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var artisticFilter = require('../artisticFilter-cb285736.cjs');



exports.ArtisticFilter = artisticFilter.ArtisticFilter;
exports.alDente = artisticFilter.alDente;
exports.athena = artisticFilter.athena;
exports.audrey = artisticFilter.audrey;
exports.aurora = artisticFilter.aurora;
exports.daguerre = artisticFilter.daguerre;
exports.eucalyptus = artisticFilter.eucalyptus;
exports.fes = artisticFilter.fes;
exports.frost = artisticFilter.frost;
exports.hairspray = artisticFilter.hairspray;
exports.hokusai = artisticFilter.hokusai;
exports.incognito = artisticFilter.incognito;
exports.linen = artisticFilter.linen;
exports.peacock = artisticFilter.peacock;
exports.primavera = artisticFilter.primavera;
exports.quartz = artisticFilter.quartz;
exports.redRock = artisticFilter.redRock;
exports.refresh = artisticFilter.refresh;
exports.sizzle = artisticFilter.sizzle;
exports.sonnet = artisticFilter.sonnet;
exports.ukulele = artisticFilter.ukulele;
exports.zorro = artisticFilter.zorro;
