// @cloudinary/url-gen/qualifiers/dither.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dither = require('../dither-9218abb1.cjs');



exports.Dither = dither.Dither;
exports.checkerboard2x1Dither = dither.checkerboard2x1Dither;
exports.circles5x5Black = dither.circles5x5Black;
exports.circles5x5White = dither.circles5x5White;
exports.circles6x6Black = dither.circles6x6Black;
exports.circles6x6White = dither.circles6x6White;
exports.circles7x7Black = dither.circles7x7Black;
exports.circles7x7White = dither.circles7x7White;
exports.halftone16x16Orthogonal = dither.halftone16x16Orthogonal;
exports.halftone4x4Angled = dither.halftone4x4Angled;
exports.halftone4x4Orthogonal = dither.halftone4x4Orthogonal;
exports.halftone6x6Angled = dither.halftone6x6Angled;
exports.halftone6x6Orthogonal = dither.halftone6x6Orthogonal;
exports.halftone8x8Angled = dither.halftone8x8Angled;
exports.halftone8x8Orthogonal = dither.halftone8x8Orthogonal;
exports.ordered2x2Dispersed = dither.ordered2x2Dispersed;
exports.ordered3x3Dispersed = dither.ordered3x3Dispersed;
exports.ordered4x4Dispersed = dither.ordered4x4Dispersed;
exports.ordered8x8Dispersed = dither.ordered8x8Dispersed;
exports.threshold1x1Nondither = dither.threshold1x1Nondither;
