// @cloudinary/url-gen/qualifiers/flag.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var flag = require('../flag-32afa2d8.cjs');
require('../FlagQualifier-0e14a6c3.cjs');
require('../QualifierValue-e770d619.cjs');
require('../Qualifier-6633a22f.cjs');
require('../QualifierModel-0923d819.cjs');
require('../unsupportedError-74070138.cjs');



exports.Flag = flag.Flag;
exports.animated = flag.animated;
exports.animatedPng = flag.animatedPng;
exports.animatedWebP = flag.animatedWebP;
exports.anyFormat = flag.anyFormat;
exports.attachment = flag.attachment;
exports.clip = flag.clip;
exports.clipEvenOdd = flag.clipEvenOdd;
exports.custom = flag.custom;
exports.forceIcc = flag.forceIcc;
exports.forceStrip = flag.forceStrip;
exports.getInfo = flag.getInfo;
exports.hlsv3 = flag.hlsv3;
exports.ignoreInitialAspectRatio = flag.ignoreInitialAspectRatio;
exports.ignoreMaskChannels = flag.ignoreMaskChannels;
exports.immutableCache = flag.immutableCache;
exports.keepAttribution = flag.keepAttribution;
exports.keepDar = flag.keepDar;
exports.keepIptc = flag.keepIptc;
exports.layerApply = flag.layerApply;
exports.lossy = flag.lossy;
exports.mono = flag.mono;
exports.noOverflow = flag.noOverflow;
exports.noStream = flag.noStream;
exports.png24 = flag.png24;
exports.png32 = flag.png32;
exports.png8 = flag.png8;
exports.preserveTransparency = flag.preserveTransparency;
exports.progressive = flag.progressive;
exports.rasterize = flag.rasterize;
exports.regionRelative = flag.regionRelative;
exports.relative = flag.relative;
exports.sanitize = flag.sanitize;
exports.splice = flag.splice;
exports.streamingAttachment = flag.streamingAttachment;
exports.stripProfile = flag.stripProfile;
exports.tiff8Lzw = flag.tiff8Lzw;
exports.tiled = flag.tiled;
exports.truncateTS = flag.truncateTS;
exports.waveform = flag.waveform;
