// @cloudinary/url-gen/qualifiers/focusOn.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var focusOn = require('../focusOn-908730ef.cjs');
var FocusOnValue = require('../FocusOnValue-b928027a.cjs');
require('../QualifierValue-e770d619.cjs');



exports.FocusOn = focusOn.FocusOn;
exports.advancedEyes = focusOn.advancedEyes;
exports.advancedFace = focusOn.advancedFace;
exports.advancedFaces = focusOn.advancedFaces;
exports.aeroplane = focusOn.aeroplane;
exports.background = focusOn.background;
exports.bicycle = focusOn.bicycle;
exports.bird = focusOn.bird;
exports.boat = focusOn.boat;
exports.bottle = focusOn.bottle;
exports.bus = focusOn.bus;
exports.car = focusOn.car;
exports.cat = focusOn.cat;
exports.chair = focusOn.chair;
exports.cow = focusOn.cow;
exports.diningtable = focusOn.diningtable;
exports.dog = focusOn.dog;
exports.face = focusOn.face;
exports.faces = focusOn.faces;
exports.horse = focusOn.horse;
exports.microwave = focusOn.microwave;
exports.motorbike = focusOn.motorbike;
exports.ocr = focusOn.ocr;
exports.person = focusOn.person;
exports.pottedplant = focusOn.pottedplant;
exports.refrigerator = focusOn.refrigerator;
exports.sheep = focusOn.sheep;
exports.sink = focusOn.sink;
exports.skateboard = focusOn.skateboard;
exports.sofa = focusOn.sofa;
exports.train = focusOn.train;
exports.tvmonitor = focusOn.tvmonitor;
exports.FocusOnValue = FocusOnValue.FocusOnValue;
