// @cloudinary/url-gen/qualifiers/format.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var format = require('../format-6def9352.cjs');
require('../FormatQualifier-ffbb8eb3.cjs');
require('../QualifierValue-e770d619.cjs');



exports.Format = format.Format;
exports.ai = format.ai;
exports.arw = format.arw;
exports.audioAac = format.audioAac;
exports.audioAiff = format.audioAiff;
exports.audioAmr = format.audioAmr;
exports.audioFlac = format.audioFlac;
exports.audioM4a = format.audioM4a;
exports.audioMp3 = format.audioMp3;
exports.audioOgg = format.audioOgg;
exports.audioOpus = format.audioOpus;
exports.audioWav = format.audioWav;
exports.auto = format.auto;
exports.avif = format.avif;
exports.bmp = format.bmp;
exports.cr2 = format.cr2;
exports.djvu = format.djvu;
exports.eps = format.eps;
exports.eps3 = format.eps3;
exports.ept = format.ept;
exports.flif = format.flif;
exports.fxb = format.fxb;
exports.gif = format.gif;
exports.glb = format.glb;
exports.gltf = format.gltf;
exports.hdp = format.hdp;
exports.heic = format.heic;
exports.heif = format.heif;
exports.ico = format.ico;
exports.indd = format.indd;
exports.jp2 = format.jp2;
exports.jpc = format.jpc;
exports.jpe = format.jpe;
exports.jpeg = format.jpeg;
exports.jpg = format.jpg;
exports.jxr = format.jxr;
exports.pdf = format.pdf;
exports.png = format.png;
exports.ps = format.ps;
exports.psd = format.psd;
exports.spd = format.spd;
exports.svg = format.svg;
exports.tga = format.tga;
exports.tif = format.tif;
exports.tiff = format.tiff;
exports.usdz = format.usdz;
exports.video3g2 = format.video3g2;
exports.video3gp = format.video3gp;
exports.videoAvi = format.videoAvi;
exports.videoFlv = format.videoFlv;
exports.videoM2ts = format.videoM2ts;
exports.videoM3u8 = format.videoM3u8;
exports.videoMkv = format.videoMkv;
exports.videoMov = format.videoMov;
exports.videoMp4 = format.videoMp4;
exports.videoMpd = format.videoMpd;
exports.videoMpeg = format.videoMpeg;
exports.videoMts = format.videoMts;
exports.videoMxf = format.videoMxf;
exports.videoOgv = format.videoOgv;
exports.videoTs = format.videoTs;
exports.videoWebm = format.videoWebm;
exports.videoWmv = format.videoWmv;
exports.wdp = format.wdp;
exports.webp = format.webp;
